src/backend/sgx/attestation.rs

```diff
@@ -0,0 +1,365 @@
+                msg.set_b_pub_key_id(false);
+                msg.set_att_key_id(akid);
+                req.set_initQuoteExReq(msg);
+                Ok(req)
+            }
+            ReqType::Quote => {
+                let akid = akid.ok_or_else(|| {
+                    Error::new(
+                        ErrorKind::Other,
+                        "Missing attestation key ID from InitQuoteExRequest",
+                    )
+                })?;
+                let report = report.ok_or_else(|| {
+                    Error::new(
+                        ErrorKind::Other,
+                        "Missing REPORT structure from GetQuoteExRequest",
+                    )
+                })?;
+                let mut msg = Request_GetQuoteExRequest::new();
+                msg.set_timeout(AESM_REQUEST_TIMEOUT);
+                msg.set_report(report.to_vec());
+                msg.set_att_key_id(akid);
+                msg.set_buf_size(SGX_QUOTE_SIZE as u32);
+                req.set_getQuoteExReq(msg);
+                Ok(req)
+            }
+        }
+    }
+
+    fn send_request(&self, req: Request, stream: &mut UnixStream) -> Result<Response, Error> {
+        // Set up writer
+        let mut buf_wrtr = vec![0u8; size_of::<u32>()];
+
+        match req.write_to_writer(&mut buf_wrtr) {
+            Ok(_) => {}
+            Err(e) => {
+                return Err(Error::new(
+                    ErrorKind::Other,
+                    format!("invalid protobuf Request: {:?}. Error: {:?}", req, e),
+                ));
+            }
+        }
+
+        let req_len = (buf_wrtr.len() - size_of::<u32>()) as u32;
+        buf_wrtr[0..size_of::<u32>()].copy_from_slice(&req_len.to_le_bytes());
+
+        // Send Request to AESM daemon
+        stream.write_all(&buf_wrtr)?;
+        stream.flush()?;
+
+        // Receive Response
+        let mut res_len_bytes = [0u8; 4];
+        stream.read_exact(&mut res_len_bytes)?;
+        let res_len = u32::from_le_bytes(res_len_bytes);
+
+        let mut res_bytes = vec![0; res_len as usize];
+        stream.read_exact(&mut res_bytes)?;
+
+        Ok(Message::parse_from_bytes(&res_bytes)?)
+    }
+}
+
+/// Gets Att Key ID
+fn get_ak_id() -> Result<Vec<u8>, Error> {
+    let mut stream = UnixStream::connect(AESM_SOCKET)?;
+
+    let r = ReqType::AkIdNum;
+    let mut req = Request::new();
+    let mut msg = Request_GetSupportedAttKeyIDNumRequest::new();
+    msg.set_timeout(AESM_REQUEST_TIMEOUT);
+    req.set_getSupportedAttKeyIDNumReq(msg);
+    let pb_msg: Response = r.send_request(req, &mut stream)?;
+
+    let res = pb_msg.get_getSupportedAttKeyIDNumRes();
+
+    if res.get_errorCode() != 0 {
+        panic!(
+            "Received error code {:?} in GetSupportedAttKeyIDNum",
+            res.get_errorCode()
+        );
+    }
+
+    let num_key_ids = res.get_att_key_id_num();
+
+    let expected_buffer_size: u32 = num_key_ids * SGX_KEY_ID_SIZE;
+    assert_eq!(1, num_key_ids);
```

Comment:
```suggestion
    debug_assert_eq!(1, num_key_ids);
```

Unless we want to panic in production

Reply:
```
    if num_key_ids != 1 {
        return Err(Error::new(
            ErrorKind::Other,
            format!("Unexpected number of key IDs: {} != 1", num_key_ids),
        ));
    }
```
I think this is a sign of not having a legit fallback, because theoretically this could very well happen in production (if components out of our control give invalid data).